python/cudf_polars/tests/dsl/test_expr.py

```diff
@@ -73,3 +73,26 @@ def test_namedexpr_repr_stable():
     b2 = expr.NamedExpr("b1", expr.Col(plc.DataType(plc.TypeId.INT8), "a"))
 
     assert repr(b1) == repr(b2)
+
+
+def test_equality_cse():
+    dt = plc.DataType(plc.TypeId.INT8)
+
+    def make_expr(n1, n2):
+        a = expr.Col(plc.DataType(plc.TypeId.INT8), n1)
+        b = expr.Col(plc.DataType(plc.TypeId.INT8), n2)
+
+        return expr.BinOp(dt, plc.binaryop.BinaryOperator.ADD, a, b)
+
+    e1 = make_expr("a", "b")
+    e2 = make_expr("a", "b")
+    e3 = make_expr("a", "c")
+
+    assert e1.children is not e2.children
+    assert e1 == e2
+    assert e1.children is e2.children
```

Comment:
Is there any other case where we coalesce common subexpressions? It happens during plan evaluation if multiple subtrees are identical (e.g. `(a == b) == (a == b)`), right? Is there anywhere else it would make sense?

Reply:
Once you use a `CachingVisitor`, CSE coalescing will happen "automatically".

We can't use it for plan evaluation because naively that requires storing everything in the tree rather than the currently live objects. If we want to use it for evaluation we'll need to have some kind of scheduling algorithm that collects refcounts of expressions and has a heuristic to decide on evaluation order and placement of the (now column-sized) temporaries.

The main reason for coalescence at this level is that if you have any tree that is actually a DAG, you want rewrites to be DAG-aware, otherwise you can (in the worst case) exponentially increase the perceived size of the object.

Reply:
I was asking because I wanted to see if there were any operations other than an equality test that might merit similar testing. I agree that a CSE analog for the plan is much more challenging and out of scope for now.

Reply:
To be more explicit: if there are any other tests that ought to be added to verify that CSE occurred after other operations, please add them. If not, feel free to mark this as resolved.

Reply:
I don't think so, dictionary insertion _might_ require it, but needn't necessarily, I think.